### ISSUE TITLE
Issues resulting from Reaper "duplicate destroy attempted"

### ISSUE THREAD

**Issue by edmorley** [heroku/hatchet#177]
In recent test runs, I've had hundreds of lines like:

```
Duplicate destoy attempted "hatchet-t-72743997b7": c1c11461-220c-414c-9be0-ce47e43454d2, status: 404, request_id: 14fad90e-9dd0-479f-b86d-03c402f6ef84,192ed965-12d2-d0b3-9a98-c2c20665888f,99c7bfc4-1f8e-858c-122f-470153e0da77
```

From my current understanding of the app reaper, these warnings are due to multiple Travis jobs all trying to delete the same apps. ie: Whilst the app reaper uses a file based mutex to prevent multiple processes on the same machine from duplicating deletion attempts, this doesn't help with CI jobs running in parallel (either multiple stacks in parallel, or CI runs from say different PRs).

The problem with these is that each duplicate destroy attempt:
1. consumes more of the scarce API rate limit
2. wastes CI runtime
3. doesn't cause the reaper to refresh `finished_hatchet_apps`, so it continues on trying to delete more apps from a now potentially stale list of deleted apps

Rather than trying to add on more complexity to avoid these duplicate destroy attempts, I was wondering whether we should instead reassess how we do app reaping in CI, given that:
* the Heroku platform has for some time supported raising the account app limit (meaning much of the purpose of the reaper is moot)
* we have the concept of "maintenance mode on means app is no longer needed by Hatchet", which offers a few different implementation choices
* in general I don't find the Hatchet default of keeping completed apps very useful - I would rather default to deleting them there and then, unless a debug option was enabled

For example, imagine a world where:
1. In CI we have the account max app limit set very high, eg 1000 (so no real chance of hitting it mid-test-run)
2. Hatchet by default deletes the app during teardown (unless the app deploy failed or a debug option was enabled, in which case it sets maintenance mode on instead). (Is it possible for Hatchet to know if an `expect` failed during the test?)
3. We add a new `hatchet destroy --old-or-completed` (or something similar), which would be run as a CI post test step (potentially just to one job in the matrix, to reduce overlap), and would mostly be a no-op given in the common case apps would only be left behind if something failed.
4. Hatchet no longer runs the reaper by default (either in CI or locally), instead only when a "max app limit reached error" is encountered, in which case it destroys completed/TTL-exceeded etc as it does currently.

The advantages of this would be:
* personal accounts no longer cluttered with hatchet apps by default (unless the "keep apps" debug mode enabled)
* much fewer API calls (the reaper no longer needs to refresh the apps list, we don't need both maintenance mode call and app delete call, plus avoiding the duplicate delete calls)
* no longer spending CI time performing the duplicate calls, or waiting on the file mutex

**Comment by edmorley** [comment]
In an ideal world, the way we handle app cleanup would:
1. Use as few API calls as possible to prevent rate limit issues
2. Consume as little CI end to end time as possible
3. Not break when Hatchet is being run in parallel across separate machines (such as in CI)
4. Leave as few Hatchet apps left around at the end of a test run as possible, to not clutter personal accounts with Hatchet apps
5. However sometimes leave apps behind to allow for debugging (either as an option, or say if the test failed to deploy or failed an expectation)
6. Handle Hatchet being aborted mid-run (eg CI job being cancelled or disappearing halfway through, which means extra cleanup needed later)
7. Prefer failing with a clear error message for suboptimal configurations (eg hitting max app limit in CI because the account limit wasn't raised) rather than adding complexity trying to work around them (several of the workarounds like exponential backoff lead to slower test runs, and using them hides the need to raise the app limit)
8. Be optimised for the common case (Herokai who is able to raise max app limit) rather than assuming app limits are always low (we can always suggest in the README that non-Herokai can file a support to have their app limit raised)

Proposed new implementation:

- We add a `--stale` option to `hatchet destroy` that limits app deletion to those older than `HATCHET_ALIVE_TTL_MINUTES`.
- We add an option that switches off the Hatchet reaper entirely during test runs, that defaults to off.
- We document the following "best practices for CI":
  - Raise the account app limit very high (eg 1000) if Herokai, otherwise to file a support ticket requesting an smaller increase if the 100 default isn't enough.
  - To enable the new the option that switches off the reaper during test runs.
  - Set `HATCHET_ALIVE_TTL_MINUTES` to a higher value than the default, so that failed deploy apps are left around for longer to aid debugging.
  - Add `hatchet destroy --stale` as a step in the CI job matrix, but **only on one of the jobs** in the matrix, to reduce contention between parallel jobs.
- We change the stale app reaper so that if it encounters an `AlreadyDeletedError` it should skip the rest of the fetched apps list, under the presumption that another machine's reaper process is handling them (such as on a different PR running in parallel), saving wasting rate limit on deletes that are mostly likely all going to be duplicates.
- On Hatchet startup, when the reaper is enabled (ie: developing locally, where we'd rather this be automatic rather than having the separate `hatchet destroy --old` step), we destroy any Hatchet apps older than `HATCHET_ALIVE_TTL_MINUTES`:
  - In the common case this will be very few apps, since (a) most will have been cleaned up during previous test teardown, (b) when developing locally the concurrency is much lower, so the number of apps left around from an aborted run (where teardown didn't run) will be lower.
  - The reaper should continue to use a file-based lock to prevent contention on the same machine (eg locally with `parallel_test` or `parallel_split_test`).
- On Hatchet app create, if the create failed, we no longer trigger an app reap (beyond that already performed at Hatchet startup), but instead provide a helpful error message suggesting: raising account app limit / reducing concurrency / using a separate test account / using `hatchet destroy --all` etc.
- On Hatchet `App.teardown` we no longer use maintenance mode at all and instead delete the app straight away, unless:
  - the app deploy failed (when `allow_failure` wasn't true)
  - an expectation failed inside the `deploy` block
  - some new debug option is enabled (either a global option or perhaps a `preserve_app: true` passed to `Hatchet::App` to allow scoping preservation to just one test when debugging?)
- We remove the concepts of:
  - `HATCHET_APP_LIMIT`. Limiting the number of Hatchet apps is not really useful when (a) Hatchet now deletes apps as it goes by default, (b) it's easy to raise app limits, (c) an account could have lots of non-Hatchet apps, so staying under `HATCHET_APP_LIMIT` is no guarantee that app creation will succeed anyway.
  - maintenance mode. Since it uses up more API calls and the problem it's trying to solve (marking which apps are complete) goes away with cleaning up by default + higher limits + potentially higher TTL etc
  - exponential backoff of waiting for apps over TTL. Adds extra complexity and slows down test runs. Better to error if exceeding app limit so users are aware and know to raise the app limit.

@schneems @dzuelke Thoughts? :-)

**Comment by schneems** [comment]
Here's my proposal broken into individual behavior changes that I think will play well together, based on our conversation and your post.

1) Mitigate the immediate issue in the reaper: We can use this throttle logic https://github.com/heroku/hatchet/blob/main/lib/hatchet/shell_throttle.rb#L11-L14 when a duplicate app deletion is detected. Then refresh the app list before trying again. This will essentially mean that the first reaper to successfully delete an app will delete multiple apps until it ends up being under the limit. The other reapers will sleep and contend for being the one to delete apps in the future.

Basically this:

> We change the stale app reaper so that if it encounters an AlreadyDeletedError it should skip the rest of the fetched apps list, under the presumption that another machine's reaper process is handling them (such as on a different PR running in parallel), saving wasting rate limit on deletes that are mostly likely all going to be duplicates.

2) Introduce behavior to record if an exception happened inside of a deploy block or not. We are already doing this here, https://github.com/heroku/hatchet/blob/a245b8c751775533948c6400f332720449b4d07d/lib/hatchet/app.rb#L99-L103 all we need to do is record the fact that it happened. We could make this general and not constrained to a specific exception.

3) If there's no recorded exception, then instead of marking an app as "maintenance mode" we can go ahead and delete it. Otherwise we'll mark it normally and let a reaper run take care of it. There's a `debug` flag passed to apps that I think either got turned into a no-op or was removed. We could consider bringing it back to optionally disable this auto-deletion if we desired.

> On Hatchet App.teardown we no longer use maintenance mode at all and instead delete the app straight away, unless:
>
>   the app deploy failed (when allow_failure wasn't true)
>    an expectation failed inside the deploy block
>    some new debug option is enabled (either a global option or perhaps a preserve_app: true passed to Hatchet::App to allow scoping preservation to just one test when debugging?)


I think that will work out of the box for most systems already in existence. It will reduce duplicate API requests when multiple reapers run, and it will decrease the need of running a reaper for the "happy path" of the test succeded and the app can be discarded.

The only case this fails is if someone is using `deploy` without block format i.e. the exception is happening somewhere we cannot observe it. I'm not sure how we want to handle that case.

4) I like the idea of a `hatchet destroy --stale` cli command that could be added to CI for those that wish to be diligent about keeping their app list clean. It's probably better for the platform as well to not have stale/unused apps around. I wish we had a platform feature of "tmp apps" that would let us set a timestamp and the platform would sweep them for us. The only thing I don't love about the CLI API, is it's a hair's breath away from accidentally typing `hatchet destroy` which I guess...isn't that catastrophic if it does happen as it only deletes hatchet apps.

**Comment by edmorley** [comment]
Fixed by #195
